hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/health/TestQuasiClosedContainerHandler.java

```diff
@@ -0,0 +1,180 @@
+        .build();
+
+    Assertions.assertFalse(quasiClosedContainerHandler.handle(request));
+    Mockito.verify(replicationManager, times(0))
+        .sendCloseContainerReplicaCommand(any(), any(), anyBoolean());
+  }
+
+  @Test
+  public void testOpenContainerReturnsFalse() {
+    ContainerInfo containerInfo = ReplicationTestUtil.createContainerInfo(
+        ratisReplicationConfig, 1, OPEN);
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicas(containerInfo.containerID(),
+            State.OPEN, 1, 2, 3);
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.EMPTY_LIST)
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    Assertions.assertFalse(quasiClosedContainerHandler.handle(request));
+    Mockito.verify(replicationManager, times(0))
+        .sendCloseContainerReplicaCommand(any(), any(), anyBoolean());
+  }
+
+  /**
+   * When a container is QUASI_CLOSED, and it has greater than 50% of its
+   * replicas in QUASI_CLOSED state with unique origin node id,
+   * the handler should send force close commands to the replica(s) with
+   * highest BCSID.
+   */
+  @Test
+  public void testQuasiClosedWithQuorumReturnsTrue() {
+    ContainerInfo containerInfo = ReplicationTestUtil.createContainerInfo(
+        ratisReplicationConfig, 1, QUASI_CLOSED);
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicas(containerInfo.containerID(),
+            State.QUASI_CLOSED, 1, 2);
+    ContainerReplica openReplica = ReplicationTestUtil.createContainerReplica(
+        containerInfo.containerID(), 3,
+        HddsProtos.NodeOperationalState.IN_SERVICE, State.OPEN);
+    containerReplicas.add(openReplica);
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.EMPTY_LIST)
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    Assertions.assertTrue(quasiClosedContainerHandler.handle(request));
+    Mockito.verify(replicationManager, times(2))
+        .sendCloseContainerReplicaCommand(any(), any(), anyBoolean());
+  }
+
+  /**
+   * The replicas are QUASI_CLOSED, but all of them have the same origin node
+   * id. Since a quorum with unique origin node ids (greater than 50% of
+   * replicas) is not formed, the handler should return false.
+   */
+  @Test
+  public void testHealthyQuasiClosedContainerReturnsFalse() {
+    ContainerInfo containerInfo = ReplicationTestUtil.createContainerInfo(
+        ratisReplicationConfig, 1, QUASI_CLOSED);
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicasWithSameOrigin(containerInfo.containerID(),
+            State.QUASI_CLOSED, 1, 2, 3);
```

Comment:
The Replica Index should always be zero for Ratis containers. Only for EC do we have the indexes 1 to N, indicating the position the container takes in the container group.

It probably doesn't cause any problems with these tests, but we should probably fix it, and just pass "0, 0, 0" rather than "1, 2, 3" for any Ratis containers as they may be other places in the code where unexpected things happen if someone reused this code later.

Reply:
Got it. Will also need to change this in the Ratis related tests for other handlers. I'll open another PR for that.

Reply:
Ah, I did not spot it then. But yea, it would be good to tidy it up there too.

Reply:
Created a jira for this https://issues.apache.org/jira/browse/HDDS-7268